File: bot/device_tiers.py
"""Per-device permission tiers for the conversational admin surface
(Server Chat's BotServer pipeline, Support Bot's new admin actions, and
device/pairing management itself) — see the "Admin control surface" plan.

Deliberately narrow and separate from the pre-existing dashboard REST
API's own flat "any paired device = desktop parity" model
(bot/dashboard/server.py's _identify_caller) — that's an existing,
already-deliberate decision this module does not touch or override.

Four ordered tiers: a device with nothing configured (`none`) gets no
admin conversational capability at all; `standard` matches the Telegram
admin bot's own tool set; `elevated` adds device/pairing management and
destructive local operations; `unrestricted` additionally skips the
per-call dangerous-tool approval prompt for run_shell/write_file, since
the tier assignment itself already represents standing consent.
"""
from __future__ import annotations

TIERS = ("none", "standard", "elevated", "unrestricted")

TIER_RANK = {tier: i for i, tier in enumerate(TIERS)}


def is_valid_tier(tier: str) -> bool:
    return tier in TIER_RANK


def can_manage(actor_tier: str, target_tier: str, *, is_self: bool) -> bool:
    """May a device at actor_tier change the tier of (or revoke) a device
    currently at target_tier? A device can never manage itself, and can
    only ever manage a STRICTLY lower-ranked device — never a peer, never
    a superior. The desktop dashboard token is not subject to this check
    at all (it's the unconditional top authority); callers should bypass
    this function entirely for a desktop caller rather than pass it in."""
    if is_self:
        return False
    return TIER_RANK.get(actor_tier, 0) > TIER_RANK.get(target_tier, 0)


def can_mint(actor_tier: str, new_tier: str) -> bool:
    """May a device at actor_tier mint a brand-new device's key at
    new_tier? A device may create a peer (same tier) but never a
    superior. The desktop dashboard token bypasses this check entirely."""
    return TIER_RANK.get(new_tier, 0) <= TIER_RANK.get(actor_tier, 0)

File: android-app/app/src/main/kotlin/com/botserver/mobile/data/DeviceTiers.kt
package com.botserver.mobile.data

/** Kotlin mirror of bot/device_tiers.py, used only to decide which
 * tier-management actions this screen *offers* — the server re-checks
 * can_manage/can_mint authoritatively on every call regardless, so a
 * stale or wrong client-side computation here can only ever hide an
 * action that would have worked, never grant one that wouldn't. */
object DeviceTiers {
    val TIERS = listOf("none", "standard", "elevated", "unrestricted")
    private val RANK = TIERS.withIndex().associate { (i, tier) -> tier to i }

    fun rank(tier: String): Int = RANK[tier] ?: 0

    /** May a device at [actorTier] change the tier of (or revoke) a
     * device currently at [targetTier]? Never itself, never a peer or
     * superior. */
    fun canManage(actorTier: String, targetTier: String, isSelf: Boolean): Boolean {
        if (isSelf) return false
        return rank(actorTier) > rank(targetTier)
    }

    /** May a device at [actorTier] mint (or retier another device to)
     * [newTier]? A peer is fine, a superior is not. */
    fun canMint(actorTier: String, newTier: String): Boolean = rank(newTier) <= rank(actorTier)

    /** Every tier this device may mint a new device at, or retier an
     * existing lower-tier device to — capped at its own tier. */
    fun mintableTiers(actorTier: String): List<String> = TIERS.filter { rank(it) <= rank(actorTier) }
}
